### ISSUE TITLE
Random ORA-08177: can't serialize access for this transaction

### ISSUE THREAD

**Issue by leolozes** [laurenz/oracle_fdw#577]
Hello,

First of all, thank you for putting the time in this project. 
We have a quite simple function in postgresql that has to copy the daily data of 2 tables to Oracle. It's usually about 300-400 rows.

The job fails randomly. Well everything is quite random, but not always. When something fails, it fails consistently. Then you change an order by, or the selected data, and everything works. It's really puzzling.

Postgres version: `PostgreSQL 12.11 on x86_64-pc-linux-gnu, 64-bit`
Oracle version:  `Oracle Database 19c Standard Edition 2 Release 19.0.0.0.0 - Production`

Both are Amazon RDS instances, so we can't control the `oracle_fdw` version:
`oracle_fdw 2.3.0, PostgreSQL 12.11, Oracle client 21.1.0.0.0, Oracle server 19.0.0.0.0`

The tables are quite simple. 

In oracle:

```sql
CREATE TABLE TMIG_BOOKING_V2 
   (	"REFERENCE" VARCHAR2(30 CHAR), 
	"CLIENT" VARCHAR2(10 CHAR), 
	"VENDOR" VARCHAR2(10 CHAR), 
	"MANAGER" VARCHAR2(10 CHAR), 
	"SYSTEM" VARCHAR2(3 CHAR), 
	"DATEFROM" DATE, 
	"DATETO" DATE, 
	"RESERVATIONSTATUS" VARCHAR2(20 CHAR), 
	"CURRENCY" VARCHAR2(3 CHAR), 
	"CLIENTREFERENCE" VARCHAR2(100 CHAR), 
	"CREATIONDATE" DATE, 
	"CREATIONUSER" VARCHAR2(10 CHAR), 
	"CANCELLATIONDATE" DATE, 
	"CONTACTPERSON" VARCHAR2(200 CHAR), 
	"CONTACTMAIL" VARCHAR2(200 CHAR), 
	"CONTACTPHONE" VARCHAR2(100 CHAR), 
	"MODIFICATIONDATE" DATE, 
	"OPERATIONALSTATUS" VARCHAR2(50 CHAR), 
	"TOTALAMOUNT" NUMBER(15,3), 
	"ORIGINALAMOUNT" NUMBER(15,3), 
	"STOREDATE" DATE, 
	"SEMCODE" VARCHAR2(250), 
	"WEBSITEURL" VARCHAR2(200), 
	"IPADDRESS" VARCHAR2(50), 
	"COUNTRY" VARCHAR2(10), 
	"REFERREDBYURL" VARCHAR2(200), 
	"REFERREDBYSYSTEM" VARCHAR2(15), 
	"MEDIUM" VARCHAR2(150), 
	"SOURCE" VARCHAR2(150), 
	 CONSTRAINT "MIG_BV2" PRIMARY KEY ("REFERENCE")
   ) ;

CREATE TABLE TMIG_BOOKING_HOTEL_V2 
   (	"REFERENCE" VARCHAR2(30), 
	"HOTELBOOKID" NUMBER(4,0), 
	"HOTEL" NUMBER(10,0), 
	"RESERVATIONSTATUS" VARCHAR2(20), 
	"DATEFROM" DATE, 
	"DATETO" DATE, 
	"UNITS" NUMBER(3,0), 
	"ROOMTYPE" VARCHAR2(5), 
	"ROOMTYPEEXTRA" VARCHAR2(100), 
	"BOARD" VARCHAR2(2), 
	"VENDOR" VARCHAR2(10), 
	"TOTALAMOUNT" NUMBER(15,3), 
	"CANCELLATIONDATE" DATE, 
	"NUMADULTS" NUMBER(3,0), 
	"NUMCHILDREN" NUMBER(3,0), 
	"NUMBABIES" NUMBER(3,0), 
	"ORIGINALAMOUNT" NUMBER(15,3), 
	"STOREDATE" DATE, 
	"TOTALROOMAMOUNT" NUMBER(15,3), 
	 CONSTRAINT "MGH_BV2" PRIMARY KEY ("REFERENCE", "HOTELBOOKID")
);
```

In postgresql:

```sql
CREATE FOREIGN TABLE tmig_booking_v2 (
	reference varchar(30) OPTIONS(key 'true') NOT NULL,
	client varchar(10) NULL,
	vendor varchar(10) NULL,
	manager varchar(10) NULL,
	"system" varchar(3) NULL,
	datefrom date NULL,
	dateto date NULL,
	reservationstatus varchar(20) NULL,
	currency varchar(3) NULL,
	clientreference varchar(100) NULL,
	creationdate timestamp NULL,
	creationuser varchar(10) NULL,
	cancellationdate timestamp NULL,
	contactperson varchar(200) NULL,
	contactmail varchar(200) NULL,
	contactphone varchar(100) NULL,
	modificationdate timestamp NULL,
	operationalstatus varchar(50) NULL,
	totalamount numeric(15, 3) NULL,
	originalamount numeric(15, 3) NULL,
	storedate timestamp NULL,
	semcode varchar(250) NULL,
	websiteurl varchar(200) NULL,
	ipaddress varchar(50) NULL,
	country varchar(10) NULL,
	referredbyurl varchar(200) NULL,
	referredbysystem varchar(15) NULL,
	medium varchar(150) NULL,
	"source" varchar(150) NULL
)
SERVER oradb
OPTIONS (table 'TMIG_BOOKING_V2');

CREATE FOREIGN TABLE tmig_booking_hotel_v2 (
	reference varchar(30) OPTIONS(key 'true') NOT NULL,
	hotelbookid numeric(4) OPTIONS(key 'true') NOT NULL,
	hotel numeric(10) NULL,
	reservationstatus varchar(20) NULL,
	datefrom date NULL,
	dateto date NULL,
	units numeric(3) NULL,
	roomtype varchar(5) NULL,
	roomtypeextra varchar(100) NULL,
	board varchar(2) NULL,
	vendor varchar(10) NULL,
	totalamount numeric(15, 3) NULL,
	cancellationdate timestamp NULL,
	numadults numeric(3) NULL,
	numchildren numeric(3) NULL,
	numbabies numeric(3) NULL,
	originalamount numeric(15, 3) NULL,
	storedate timestamp NULL,
	totalroomamount numeric(15, 3) NULL
)
SERVER oradb
OPTIONS (table 'TMIG_BOOKING_HOTEL_V2');
```

And the function to copy the data:

```sql
create or replace function f_booking_v2_to_v1()
 returns void
 language plpgsql
as $function$
declare
  _reference text;
  _all_references text[];
  _created_references text[];
  _updated_references text[];
  --
  message text;
  detail text;
  hint text;
  --
begin
  --
  raise info 'Start booking copy process';
  --
  _created_references := ARRAY(select reference from tbkh_booking tb where creationdate::date = current_date - 1 order by reference);
  _updated_references := ARRAY(select reference from tbkh_booking tb where modificationdate::date = current_date - 1 order by reference);
  --
  _all_references = ARRAY( SELECT DISTINCT reference FROM unnest(array_cat(_created_references, _updated_references)) as a(reference));
  --
  delete from tmig_booking_V2 where reference = any(_all_references);
  --
  delete from tmig_booking_hotel_V2 where reference = any(_all_references);
  --
  foreach _reference in array _all_references
  loop
    begin
      insert into tmig_booking_V2 (
        reference, client, vendor, manager, "system", datefrom, dateto, reservationstatus, currency, clientreference,
        creationdate, creationuser, cancellationdate, contactperson, contactmail, contactphone, modificationdate, 
        operationalstatus, totalamount, originalamount, storedate, semcode, websiteurl, ipaddress, country, referredbyurl, referredbysystem, medium, "source"
      )
      select reference, clientcode, vendorcode , manager, "system", datefrom , dateto, reservationstatus, currency, clientreference,
             creationdate, creationuser, cancellationdate, contactperson, contactmail, contactphone, modificationdate, 
             operationalstatus, totalamount, originalamount, storedate, utmcampaign, websiteurl, ipaddress, country, referredbyurl, referredbysystem, utmmedium, utmsource
      from tbkh_booking tb
      where tb.reference = _reference;
      --
      exception when others then
      get stacked diagnostics message = MESSAGE_TEXT, detail = PG_EXCEPTION_DETAIL, hint = PG_EXCEPTION_HINT;
      raise notice 'Error with reference % % % %', _reference, message, detail, hint;
    end;
  end loop;
  --
  foreach _reference in array _all_references
  loop
    begin
      --
      --raise info 'Start copying hotel booking % %', _reference, idx;   
      --
      insert into tmig_booking_hotel_V2 (
        reference, hotelbookid, hotel, reservationstatus, datefrom, dateto, units, roomtype, roomtypeextra, 
        board, vendor, totalamount, cancellationdate, numadults, numchildren, numbabies, originalamount, storedate
      )
      select reference, roomid, hotelcode, reservationstatus, datefrom, dateto, 1, SPLIT_PART(roomcode, '#', 1) roomtype, SPLIT_PART(roomcode, '#', 2) roomtypeextra, 
             mealplancode, vendorcode, totalamount, cancellationdate, adults, children, babies, originalamount, storedate
      from tbkh_booking_room tbr
      where reference =_reference;
      --
      exception when others then
      get stacked diagnostics message = MESSAGE_TEXT, detail = PG_EXCEPTION_DETAIL, hint = PG_EXCEPTION_HINT;
      raise notice 'Error with reference % % % %', _reference, message, detail, hint;
      --
    end;
  end loop;
  --
  raise info 'End booking copy process successfuly';
  --
end;
$function$
;

```

As you can see, there's no black magic here.
The job worked for 3 days straight, then failed with ORA-08177 repeatedly. 
When it fails, it fails with the same error subsequently, so it's reproducible. 
But if I change the dataset, or the order of the first select, it works. If I run it only with the data that failed, it works. 
So it's not a data problem, and it's not a communication problem. I'm really struggling to understand what is going on here.

I'd appreciate any tips or if I'm doing something fundamentally wrong, let me know :)

**Comment by laurenz** [comment]
You did everything right; the guilty part is Oracle. They just did a very bad job of implementing `SERIALIZABLE` transactions. If there is any concurrent DML activity, or even if there is an index page split during an `INSERT`, you will get these serialization errors.

The problem that I have with oracle_fdw is that it needs serializable transactions for correct operation: if you run a more complicated query involving a foreign table, it could be that that results in several Oracle queries. To make sure that the result of the query is consistent, `SERIALIZABLE` (with all its unpleasant side effects) is the only way.

Now you are not the first one to encounter that problem, so there is a way out: you can set the option `isolation_level = 'read_committed'` on the foreign server. Then that isolation level will be used. As long as you are running simple queries or don't mind inconsistent query results, you can do that, and it will free you from these pesky serialization errors.

**Comment by leolozes** [comment]
Hi @laurenz! Thanks for the quick reply.

We did try another connection with `read_commited`:

`CREATE SERVER oradb2 FOREIGN DATA WRAPPER oracle_fdw OPTIONS (dbserver '//htt-pro-oracle-01.cxudewfcsnjo.eu-west-1.rds.amazonaws.com:1521/HTTPRO', isolation_level 'read_committed');`

I'm pretty sure we already tried this, since it's in the docs, but we'll check again and try running the job with this connection instead.
